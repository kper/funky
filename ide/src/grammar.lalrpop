use std::str::FromStr;
use crate::ir::ast::*;

grammar;

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

pub Program : Program = {
   <functions:Function*> => Program {
       functions: functions
   } 
};

pub Function : Function = {
    "define" <name:Id> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Id> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    }
};

pub Instruction : Instruction = {
    "BLOCK" <n:Id> => Instruction::Block(n),
    <dest:Reg> "=" <src:Reg> => Instruction::Assign(dest, src),
    <dest:Reg> "=" <n:Num> => Instruction::Const(dest, n),
    <dest:Reg> "=" "UNKNOWN" => Instruction::Unknown(dest),
    <dest:Reg> "=" "op" <src:Reg> => Instruction::Unop(dest, src),
    <dest:Reg> "=" <src1:Reg> "op" <src2:Reg> => Instruction::BinOp(dest, src1, src2),
    <dest:Reg> "=" "phi" <src1:Reg> <src2:Reg> => Instruction::Phi(dest, src1, src2),
    <dest:Reg> "=" "STORE" <src:Reg> "TO" <offset:Reg> "ALIGN" <align:Num> <size:Num> => Instruction::Store,
    "GOTO" <n:Id> => Instruction::Jump(n),
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> "ELSE" "GOTO" <n2:Id> => {
        Instruction::Conditional(reg, vec![n, n2])
    },
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> => {
        Instruction::Conditional(reg, vec![n])
    },
    "TABLE GOTO" <n:Id*> "ELSE" "GOTO" <n2:Id> => {
        let mut n = n.clone();
        n.push(n2);
        Instruction::Table(n)
    },
    "KILL" <reg:Reg> => Instruction::Kill(reg),
    "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "RETURN" <n:Reg*> ";" => {
        Instruction::Return(n)
    },
    "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, vec![]),
    <dest:Reg*> "<-" "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, dest),
};

Id : String = {
    <n:Num> => (n as usize).to_string()
};

pub Reg: String = {
    "%" <n:Num> => format!("%{}", n),
    "%" <n:Name> => format!("%{}", n),
};

pub Num: f64 = <s:r"[+-]?[0-9]*[.,]?[0-9]+"> => f64::from_str(s).unwrap();
pub Name: String = <s:r"[a-zA-Z]+"> => s.to_string();