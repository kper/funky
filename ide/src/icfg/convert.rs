/// This module is responsible to parse
/// the webassembly AST to a graph

use crate::counter::Counter;
use crate::icfg::graph::Graph;
use crate::ssa::ast::Instruction;
use anyhow::{bail, Context, Result};

use log::debug;

use crate::icfg::graph::Fact;
use std::collections::HashMap;

use crate::ssa::ast::Program;

#[derive(Debug)]
pub struct Convert {
    block_counter: Counter,
    functions: HashMap<String, usize>, // function name to fact id
    registration_returns: HashMap<String, (Vec<Fact>, Vec<String>)>, // function name to (all facts + dest regs)
}

impl Convert {
    pub fn new() -> Self {
        Self {
            block_counter: Counter::default(),
            functions: HashMap::new(),
            registration_returns: HashMap::new(),
        }
    }

    pub fn visit(&mut self, prog: Program) -> Result<Graph> {
        debug!("Convert intermediate repr to graph");

        let mut graph = Graph::new();

        for function in prog.functions.iter() {
            debug!("Init function {}", function.name);
            graph.init_function(function);
            //graph.add_row(&function.name, format!("init {}", function.name), &mut Vec::new())?;
        }

        //debug!("graph {:#?}", graph);

        for function in prog.functions.iter() {
            debug!("Creating graph from function {}", function.name);

            // New function
            graph.epoch.push();

            let mut iterator =
                InstructionIterator::new(function.instructions.iter().collect::<Vec<_>>());

            for instruction in &mut iterator {
                match instruction {
                    Instruction::Const(reg, _val) => {
                        debug!("Adding const");

                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;

                        graph.add_var(&function.name, &reg)?;

                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Assign(dest, src) => {
                        debug!("Assignment");

                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;

                        graph.add_assignment(&function.name, &dest, &src)?;

                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Unop(dest, src) => {
                        debug!("Unop");
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;
                        graph.add_unop(&function.name, &dest, &src)?;
                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::BinOp(dest, src1, src2) => {
                        debug!("Binop");
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;

                        graph.add_binop(&function.name, &dest, &src1, &src2)?;
                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Kill(dest) => {
                        debug!("Kill");
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;
                        graph.kill_var(&function.name, &dest)?;
                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Call(name, _params, dest_regs) => {
                        graph.add_empty_vars(&function.name, &dest_regs)?;
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;
                        let meeting_facts = graph.add_call_to_return(
                            &function.name,
                            format!("Return from {}", name),
                        )?;

                        // Expect a return edge from `name` with function.name to the meeting facts
                        self.registration_returns
                            .insert(name.clone(), (meeting_facts, dest_regs.clone()));
                    }
                    _ => {}
                }
            }
            
            // function ended
            graph.epoch.pop();
        }

        for function in prog.functions.iter() {
            let mut iterator =
                InstructionIterator::new(function.instructions.iter().collect::<Vec<_>>());

            for instruction in &mut iterator {
                match instruction {
                    Instruction::Call(name, params, _dest_regs) => {
                        debug!("Call {}", name);

                        let lookup_function = prog
                            .functions
                            .iter()
                            .find(|x| &x.name == name)
                            .context("Function not found")?;

                        graph.add_call(&function.name, &lookup_function, name, params)?;
                    }
                    _ => {}
                }
            }
        }

        for (goal_function, (meeting_facts, dest_regs)) in self.registration_returns.drain() {
            if let Some((goal_function_name, goal_function_results_len, goal_function_last_facts)) = 
                graph.functions.get(&goal_function).map(|x| (&x.name, x.results_len, x.last_facts.clone())) {

                if goal_function_results_len != dest_regs.len() {
                    bail!("Mismatch results with call of {}.\nExpected results: {}\nActual results: {}", 
                        goal_function_name, 
                        goal_function_results_len, 
                        dest_regs.len());
                }

                graph.add_return(
                    &goal_function_last_facts,
                    meeting_facts,
                    &dest_regs,
                )?;
            }
        }

        return Ok(graph);
    }
}

/// Highlevel iterator, that has the
/// ability to jump to blocks
struct InstructionIterator<'a> {
    instructions: Vec<&'a Instruction>,
    current: usize,
    blocks: HashMap<&'a String, usize>, // BlockId to index in instructions
}

impl<'a> InstructionIterator<'a> {
    pub fn new(instructions: Vec<&'a Instruction>) -> Self {
        let mut v = Self {
            instructions,
            current: 0,
            blocks: HashMap::default(),
        };

        v.create_block_jump_list();

        v
    }

    fn create_block_jump_list(&mut self) {
        for (index, block) in self
            .instructions
            .iter()
            .enumerate()
            .filter(|(_id, x)| matches!(x, Instruction::Block(_)))
        {
            match block {
                Instruction::Block(id) => {
                    self.blocks.insert(id, index);
                }
                _ => {
                    panic!("Only expecting blocks");
                }
            }
        }
    }

    pub fn jump_to(&mut self, block_id: &String) -> Result<()> {
        debug!("Jump to block {}", block_id);
        if let Some(index) = self.blocks.get(block_id) {
            self.current = *index;
            return Ok(());
        } else {
            bail!("Block {} does not exist", block_id);
        }
    }

    #[allow(dead_code)]
    pub fn peek(&self) -> Option<&'a Instruction> {
        self.instructions.get(self.current).map(|x| *x)
    }
}

impl<'a> std::iter::Iterator for &mut InstructionIterator<'a> {
    type Item = &'a Instruction;

    fn next(&mut self) -> Option<Self::Item> {
        let item = self.instructions.get(self.current);
        self.current += 1;
        let item = item.map(|x| *x);

        match item {
            Some(Instruction::Block(_)) => {
                return self.next();
            }
            _ => {}
        }

        if let Some(&Instruction::Jump(ref id)) = item {
            if self.jump_to(id).is_ok() {
                self.next()
            } else {
                debug!("Block not found, therefore ending");
                None
            }
        } else {
            item
        }
    }
}
