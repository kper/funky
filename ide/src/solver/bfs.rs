use crate::icfg::graph2::Graph;
use crate::solver::*;
use std::collections::VecDeque;

use log::debug;

pub struct BFS;

impl GraphReachability for BFS {
    fn all_sinks(&mut self, graph: &mut Graph, req: Request) -> Vec<Taint> {
        let mut queue: VecDeque<usize> = VecDeque::new();
        let mut seen = Vec::new();

        if let Some(start) = graph.query(&req) {
            debug!("Adding start {:#?}", start);
            queue.push_back(start.id);
        } else {
            return Vec::new();
        }

        while let Some(node) = queue.pop_front() {
            debug!("Popping node {}", node);
            seen.push(node);
            for child in graph.get_neighbours(node) {
                debug!("Adding child {:#?}", graph.query_by_fact_id(child));

                if !seen.contains(&child) {
                    debug!("queue child {}", child);
                    queue.push_back(child);
                }
            }
        }

        seen.into_iter()
            .map(|x| graph.query_by_fact_id(x).unwrap())
            .map(|x| Taint {
                from: req.variable.clone(),
                from_function: req.function.clone(),
                from_pc: req.pc,
                to: x.belongs_to_var.clone(),
                to_pc: x.pc,
                to_function: x.function.clone(),
            })
            .collect()
    }
}
