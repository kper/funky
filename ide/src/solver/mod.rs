use crate::icfg::graph2::Graph;

pub mod bfs;

type PC = usize;

pub struct IfdsSolver<T>
where
    T: GraphReachability,
{
    graph_reachability: T,
}

#[derive(Debug)]
pub struct Taint {
    pub from: String,
    pub from_function: String,
    pub from_pc: PC,
    pub to: String,
    pub to_pc: PC,
    pub to_function: String,
}

#[derive(Debug, Clone)]
pub struct Request {
    pub variable: String,
    pub function: String,
    pub pc: PC,
}

pub trait Solver {
    /// Return all sinks of the `req`
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
    /// Ask a query the solver
    fn ask(&mut self, graph: &mut Graph, req: &Request, response: &Request) -> bool;

    /// Ask a query the solver, but doesn't care about the program counter
    fn fast_ask(
        &mut self,
        graph: &mut Graph,
        req: &Request,
        response_var: &String,
        response_function: &String,
    ) -> bool;
}

impl<T> IfdsSolver<T>
where
    T: GraphReachability,
{
    pub fn new(algorithm: T) -> Self {
        Self {
            graph_reachability: algorithm,
        }
    }
}

impl<T> Solver for IfdsSolver<T>
where
    T: GraphReachability,
{
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint> {
        self.graph_reachability.all_sinks(graph, req)
    }

    fn ask(&mut self, graph: &mut Graph, req: &Request, response: &Request) -> bool {
        // TODO optimize
        // There is no need to run the whole graph reachability algorithm
        let sinks = self.graph_reachability.all_sinks(graph, req);

        sinks
            .iter()
            .find(|x| {
                x.to == response.variable
                    && x.to_pc == response.pc
                    && x.to_function == response.function
            })
            .is_some()
    }

    fn fast_ask(
        &mut self,
        graph: &mut Graph,
        req: &Request,
        response_var: &String,
        response_function: &String,
    ) -> bool {
        // TODO optimize
        // There is no need to run the whole graph reachability algorithm
        let sinks = self.graph_reachability.all_sinks(graph, req);

        sinks
            .iter()
            .find(|x| &x.to == response_var && &x.to_function == response_function)
            .is_some()
    }
}

pub trait GraphReachability {
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
}
