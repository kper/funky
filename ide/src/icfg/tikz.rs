use crate::icfg::graph2::{Edge, Graph};
use log::debug;

pub const TAU: usize = 1;

pub fn render_to(graph: &Graph) -> String {
    let mut str_vars = String::new();

    let mut index = 0;
    let mut functions: Vec<_> = graph.functions.iter().map(|(_, f)| f).collect();

    functions.sort_by(|a, b| b.name.cmp(&a.name));

    for function in functions {
        let function_name = &function.name;
        debug!("Drawing function {}", function_name);

        let facts = graph.facts.iter().filter(|x| &x.function == function_name);
        let notes = graph.notes.iter().filter(|x| &x.function == function_name);

        for fact in facts {
            debug!("Drawing fact");

            str_vars.push_str(&format!(
                "\\node[circle,fill,inner sep=1pt,label=left:{}] ({}) at ({}, {}) {{}};\n",
                fact.belongs_to_var.replace("%", "\\%"),
                fact.id,
                index + fact.track,
                fact.pc,
            ));
        }

        for note in notes {
            debug!("Drawing note");

            str_vars.push_str(&format!(
                "\\node[font=\\tiny] (note_{}) at ({}, {}) {{{}}};\n",
                note.id,
                index as f64 - 1.5,   //x
                note.pc as f64 - 0.5, //y
                note.note.replace("%", "").replace("\"", "").escape_debug(),
            ));
        }

        index += function.definitions + TAU + 1;
    }

    for edge in graph.edges.iter() {
        match edge {
            Edge::Normal { from, to , curved } => {
                if *curved {
                    str_vars.push_str(&format!("\t\t\\path[->, bend right] ({}) edge ({});\n", from.id, to.id));
                }
                else {
                    str_vars.push_str(&format!("\t\t\\path[->] ({}) edge ({});\n", from.id, to.id));
                }
            }
            Edge::Call { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, green] ({}) [bend left] edge node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
            Edge::CallToReturn { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->] ({}) edge node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
            Edge::Return { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend right] edge  node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
        }
    }

    template(str_vars)
}

fn template(inject: String) -> String {
    format!(
        "   \\documentclass{{article}}
            \\usepackage{{pgf, tikz}}
            \\usetikzlibrary{{arrows, automata}}
            \\begin{{document}}
                \\begin{{tikzpicture}}
                    [node distance=3cm]
                    \\tikzstyle{{every state}}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    {}

                \\end{{tikzpicture}}
            \\end{{document}}",
        inject
    )
}
