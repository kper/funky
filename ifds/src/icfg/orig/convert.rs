/// This module is responsible to parse
/// the webassembly AST to a graph
use crate::icfg::graph::*;
use crate::icfg::state::State;
use crate::ir::ast::Function as AstFunction;
use crate::ir::ast::Instruction;

use crate::solver::Request;
use anyhow::{Context, Result};
use std::collections::VecDeque;

use log::debug;

use crate::ir::ast::Program;

use crate::icfg::naive::convert::Convert as NaiveConvert;
use std::collections::HashMap;

type CallerFunction = String;
type PC = usize;
type Function = String;
type CallResolver = HashMap<Function, Vec<(CallerFunction, PC, Vec<String>)>>;
pub(crate) struct Ctx<'a> {
    pub graph: &'a mut Graph,
    pub new_graph: &'a mut Graph,
    pub state: &'a mut State,
    pub call_resolver: &'a CallResolver,
}

/// Central datastructure for the computation of the IFDS problem.
#[derive(Debug, Default)]
pub struct OriginalConvert;

impl OriginalConvert {
    /// Computes a graph by a given program and `req` ([`Request`]).
    /// The `variable` in `req` doesn't matter. It only matters the `function` and `pc`.
    pub fn visit(&mut self, prog: &Program, req: &Request) -> Result<(Graph, State)> {
        let mut fact_gen = NaiveConvert::default();
        let (mut graph, mut state, call_resolver) = fact_gen
            .visit(prog)
            .context("Naive fact generation failed")?;

        let mut new_graph = Graph::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            new_graph: &mut new_graph,
            call_resolver: &call_resolver,
        };

        self.tabulate(&mut ctx, prog, req)?;

        Ok((new_graph, state))
    }

    fn tabulate<'a>(&mut self, ctx: &mut Ctx<'a>, prog: &Program, req: &Request) -> Result<()> {
        debug!("Convert intermediate repr to graph");

        let function = prog
            .functions
            .iter()
            .find(|x| x.name == req.function)
            .context("Cannot find function")?;

        let facts = ctx
            .state
            .get_facts_at(&function.name, req.pc)?
            .cloned()
            .collect::<Vec<_>>();

        let mut path_edge = Vec::new();
        let mut worklist = VecDeque::new();
        let mut summary_edge = Vec::new();
        let mut normal_flows_debug = Vec::new();

        let init = facts.get(0).unwrap().clone();

        // self loop for taut
        self.propagate(
            &mut ctx.new_graph,
            &mut path_edge,
            &mut worklist,
            Edge::Path {
                from: init.clone(),
                to: init.clone(),
            },
        )?;

        self.forward(
            &prog,
            &function,
            &mut path_edge,
            &mut worklist,
            &mut summary_edge,
            &mut normal_flows_debug,
            ctx,
            req.pc,
        )?;

        Ok(())
    }

    fn propagate(
        &self,
        graph: &mut Graph,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        e: Edge,
    ) -> Result<()> {
        let from = e.get_from();
        let to = e.to();

        let f = path_edge
            .iter()
            .find(|x| x.get_from() == from && x.to() == to);

        if f.is_none() {
            debug!("Propagate {:#?}", e);
            graph.edges.push(e.clone());
            path_edge.push(e.clone());
            worklist.push_back(e);
        }

        Ok(())
    }

    fn forward<'a>(
        &mut self,
        program: &Program,
        _function: &AstFunction,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        summary: &mut Vec<Edge>,
        normal_flows_debug: &mut Vec<Edge>,
        ctx: &mut Ctx<'a>,
        start_pc: usize,
    ) -> Result<()> {
        while let Some(edge) = worklist.pop_front() {
            debug!("Popping edge from worklist {:#?}", edge);

            assert!(
                matches!(edge, Edge::Path { .. }),
                "Edge in the worklist has wrong type"
            );

            let d1 = edge.get_from();
            let d2 = edge.to();

            let pc = edge.to().next_pc;
            debug!("Instruction pointer is {}", pc);

            let instructions = &program
                .functions
                .iter()
                .find(|x| x.name == d2.function)
                .context("Cannot find function")?
                .instructions;
            let n = instructions.get(pc);
            debug!("=> Instruction {:?}", n);

            if let Some(n) = n {
                match n {
                    Instruction::Call(callee, _params, _dest) => {
                        // handle_call

                        let caller_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d1.function)
                            .unwrap();

                        let callee_function = program
                            .functions
                            .iter()
                            .find(|x| &x.name == callee)
                            .unwrap();

                        let flow_edges = ctx.graph.edges.iter().filter(|x| {
                            x.get_from().function == caller_function.name
                                && x.get_from().belongs_to_var == d2.belongs_to_var
                                && x.get_from().next_pc == d2.next_pc
                                && x.to().function == callee_function.name
                                && x.to().next_pc == 0
                        });

                        for f in flow_edges.into_iter() {
                            let to = f.to();
                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: to.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }

                        let flow_edges = ctx
                            .graph
                            .edges
                            .iter()
                            .chain(summary.iter())
                            .filter(|x| {
                                x.get_from().function == caller_function.name
                                    && x.get_from().belongs_to_var == d2.belongs_to_var
                                    && x.get_from().next_pc == d2.next_pc
                                    && x.to().function == caller_function.name
                                    //&& x.to().belongs_to_var == x.get_from().belongs_to_var
                                    && x.to().next_pc == x.get_from().next_pc + 1
                            })
                            .map(|x| x.to());

                        for to in flow_edges.into_iter() {
                            assert_eq!(d1.function, to.function);

                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                    _ => {
                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();

                        let flow_edges =
                            ctx.graph
                                .edges
                                .iter()
                                .filter(|x| x.is_normal())
                                .filter(|x| {
                                    x.get_from().function == new_function.name
                                        && x.get_from().belongs_to_var == d2.belongs_to_var
                                        && x.get_from().next_pc == d2.next_pc
                                });

                        for f in flow_edges.into_iter() {
                            debug!("Normal flow {:#?}", f);
                            let to = f.to();

                            normal_flows_debug.push(f.clone());

                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                }
            } else {
                // end procedure

                if let Some(callers) = ctx.call_resolver.get(&d1.function) {
                    let callers = callers
                        .iter()
                        .map(|(caller, _pc, _dests)| (caller, &d1.function));

                    for (caller, callee) in callers {
                        assert_eq!(callee, &d1.function);

                        let d4 = ctx
                            .graph
                            .edges
                            .iter()
                            .filter(|x| x.is_call())
                            .filter(|x| {
                                &x.get_from().function == caller
                                    && x.to().function == d1.function
                                    && x.to().belongs_to_var == d1.belongs_to_var
                                    && x.to().next_pc == d1.next_pc
                            })
                            .map(|x| x.get_from());

                        for d4 in d4.into_iter() {
                            let d5 = ctx
                                .graph
                                .edges
                                .iter()
                                .filter(|x| x.is_return())
                                .filter(|x| {
                                    x.get_from().belongs_to_var == d2.belongs_to_var
                                    && x.get_from().function == d2.function
                                    && x.get_from().next_pc == d2.next_pc - 1 //because applied before
                                    && &x.to().function == &d4.function
                                    // && x.to().next_pc == pc + 1
                                })
                                .map(|x| x.to());

                            for d5 in d5 {
                                if summary
                                    .iter()
                                    .find(|x| {
                                        &x.get_from().function == caller
                                            && x.get_from().belongs_to_var == d4.belongs_to_var
                                            && &x.to().function == caller
                                            && x.to().next_pc == x.get_from().next_pc + 1
                                            && x.to().belongs_to_var == d5.belongs_to_var
                                    })
                                    .is_none()
                                {
                                    let ret = d4.clone();
                                    summary.push(Edge::Summary {
                                        from: d4.clone(),
                                        to: ret.clone(),
                                    });

                                    let edges = ctx
                                        .graph
                                        .edges
                                        .iter()
                                        .filter(|x| {
                                            &x.get_from().function == caller
                                                && &x.to().function == caller
                                                && x.get_from().next_pc == start_pc
                                                && x.to().belongs_to_var == d4.belongs_to_var
                                                && x.to().next_pc == d4.next_pc
                                        })
                                        .map(|x| x.get_from());

                                    for d3 in edges {
                                        let mut ret = ret.clone();
                                        ret.belongs_to_var = d5.belongs_to_var.clone();

                                        let init = ctx
                                            .state
                                            .get_facts_at(caller, 0)?
                                            .find(|x| x.belongs_to_var == d3.belongs_to_var)
                                            .context("Cannot find fact")?
                                            .clone();

                                        self.propagate(
                                            &mut ctx.new_graph,
                                            path_edge,
                                            worklist,
                                            Edge::Path {
                                                from: init,
                                                to: ret,
                                            },
                                        )?;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        //graph.edges.extend_from_slice(&path_edge);
        //graph.edges.extend_from_slice(&normal_flows_debug);
        //graph.edges.extend_from_slice(&summary_edge);

        Ok(())
    }
}
