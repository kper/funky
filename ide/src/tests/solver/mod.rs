use crate::solver::bfs::*;
use crate::solver::*;

use crate::icfg::convert::Convert;
use crate::icfg::tikz::render_to;
use insta::assert_snapshot;

use crate::grammar::*;

macro_rules! ir {
    ($name:expr, $ir:expr) => {{
        let mut convert = Convert::new();

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let graph = convert.visit(&prog).unwrap();

        let output = render_to(&graph);

        assert_snapshot!(format!("{}_dot", $name), output);

        graph
    }};
}

fn vars(sink: &Vec<Taint>) -> Vec<String> {
    let mut touched_vars: Vec<_> = sink.iter().map(|x| x.to.clone()).collect();
    touched_vars.sort_unstable();
    touched_vars.dedup();

    touched_vars
}

fn functions(sink: &Vec<Taint>) -> Vec<String> {
    let mut touched_vars: Vec<_> = sink.iter().map(|x| x.to_function.clone()).collect();
    touched_vars.sort_unstable();
    touched_vars.dedup();

    touched_vars
}

#[test]
fn test_bfs_reachability_simple() {
    let mut solver = IfdsSolver::new(BFS);

    let mut graph = ir!(
        "bfs_simple",
        "
        define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = %0
         };
    "
    );

    let sinks = solver.all_sinks(
        &mut graph,
        &Request {
            variable: "%0".to_string(),
            function: "test".to_string(),
            pc: 1,
        },
    );

    assert_eq!(3, sinks.len());

    let touched_vars = vars(&sinks);
    assert_eq!(2, touched_vars.len());

    let touched_funcs = functions(&sinks);
    assert_eq!(1, touched_funcs.len());
}

#[test]
fn test_bfs_reachability_call() {
    let mut solver = IfdsSolver::new(BFS);

    let mut graph = ir!(
        "bfs_call",
        "
        define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = %0
            %1 <- CALL mytest(%0)
        };
        define mytest (param %2) (result 1) (define %2 %3) {
            %3 = %2
            RETURN %3;
        };
    "
    );

    let sinks = solver.all_sinks(
        &mut graph,
        &Request {
            variable: "%0".to_string(),
            function: "test".to_string(),
            pc: 1,
        },
    );

    assert_eq!(10, sinks.len());

    let touched_vars = vars(&sinks);
    assert_eq!(4, touched_vars.len());

    let touched_funcs = functions(&sinks);
    assert_eq!(2, touched_funcs.len());
}

#[test]
fn test_bfs_functions() {
    let mut solver = IfdsSolver::new(BFS);

    let mut graph = ir!("bfs_functions", 
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 = %1
        };
        define mytest (param %0) (result 1) (define %0) {
            RETURN %0;
        };"
    );

    let sinks = solver.all_sinks(
        &mut graph,
        &Request {
            variable: "%0".to_string(),
            function: "test".to_string(),
            pc: 1,
        },
    );

    assert!(sinks.iter().find(|x| x.to_pc == 1 && x.to_function == "test".to_string()).is_some());
    assert!(sinks.iter().find(|x| x.to_pc == 2 && x.to_function == "test".to_string()).is_some());
    assert!(sinks.iter().find(|x| x.to_pc == 3 && x.to_function == "test".to_string()).is_some());
    assert!(sinks.iter().find(|x| x.to_pc == 1 && x.to_function == "mytest".to_string()).is_some());
}