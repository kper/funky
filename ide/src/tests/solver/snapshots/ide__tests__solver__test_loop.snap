---
source: src/tests/solver/mod.rs
expression: "format!(\"{:#?}\", sinks)"

---
[
    Taint {
        variable: "taut",
        function: "main",
        pc: 0,
    },
    Taint {
        variable: "%0",
        function: "main",
        pc: 2,
    },
    Taint {
        variable: "%0",
        function: "main",
        pc: 0,
    },
    Taint {
        variable: "%0",
        function: "main",
        pc: 1,
    },
]
